File: naiveserver/src/main/java/com/lightbend/naive/modelserver/ModelProcessor.java
package com.lightbend.naive.modelserver;

import com.lightbend.model.CurrentModelDescriptor;
import com.lightbend.model.DataConverter;
import com.lightbend.model.ModelWithDescriptor;
import com.lightbend.naive.modelserver.queriablestate.ModelServingInfo;
import com.lightbend.naive.modelserver.store.StoreState;
import org.apache.kafka.streams.processor.AbstractProcessor;
import org.apache.kafka.streams.processor.ProcessorContext;

import java.util.Optional;

/**
 * Created by boris on 7/12/17.
 */

public class ModelProcessor extends AbstractProcessor<byte[], byte[]> {

    private StoreState modelStore;

    @Override
    public void process(byte[] key, byte[] value) {

        Optional<CurrentModelDescriptor> descriptor = DataConverter.convertModel(value);
        if(!descriptor.isPresent()){
            return;                                                 // Bad record
        }
        Optional<ModelWithDescriptor> modelWithDescriptor = DataConverter.convertModel(descriptor);
        if(modelWithDescriptor.isPresent()){
           modelStore.setNewModel(modelWithDescriptor.get().getModel());
           modelStore.setNewServingInfo(new ModelServingInfo(modelWithDescriptor.get().getDescriptor().getName(),
                    modelWithDescriptor.get().getDescriptor().getDescription(), 0));
           return;
        }
    }

    @Override
    public void init(ProcessorContext context) {
        modelStore = StoreState.getInstance();
    }
}

File: kafkaStreamsModelServerInMemoryStore/src/main/scala/com/lightbend/scala/naive/modelserver/ModelProcessor.scala
package com.lightbend.scala.naive.modelserver

import com.lightbend.scala.modelServer.model.{ModelToServe, ModelToServeStats, ModelWithDescriptor}
import com.lightbend.scala.naive.modelserver.store.StoreState
import org.apache.kafka.streams.processor.{AbstractProcessor, ProcessorContext, ProcessorSupplier}

import scala.util.Success

class ModelProcessor extends AbstractProcessor[Array[Byte], Array[Byte]]{

  private var modelStore: StoreState = null

  override def process (key: Array[Byte], value: Array[Byte] ): Unit = {

    ModelToServe.fromByteArray(value) match {
      case Success(descriptor) => {
        ModelWithDescriptor.fromModelToServe(descriptor) match {
          case Success(modelWithDescriptor) => {
            modelStore.newModel = Some(modelWithDescriptor.model)
            modelStore.newState = Some(ModelToServeStats(descriptor))
          }
          case _ => // ignore
        }
      }
      case _ => // ignore
    }
  }

  override def init(context: ProcessorContext): Unit = {
    modelStore = StoreState()
  }
}
